### PR TITLE
Add Kourier LoadBalancerIP Config

### LINKED ISSUE #1290: Specify Kourier Load Balancer IP
**Problem**
There's no declarative way of specifying the Load Balancer IP that Kourier should use.

**[Persona:](https://github.com/knative/eventing/blob/main/docs/personas.md)**
System Operator

**Exit Criteria**
There should be a way of specifying Kourier's Load Balancer IP using the Operator.
https://github.com/knative-sandbox/net-kourier/issues/747

### PR DESCRIPTION
**Pull request by theomessin**
Fixes #1290

## Proposed Changes

* Adds new `service-load-balancer-ip` config for the Kourier Gateway.
* This sets Kourier's Service `loadBalancerIP` to the configured value.
* Several Cloud Providers use this setting to specify the IP address that a service gets assigned:

> Some cloud providers allow you to specify the loadBalancerIP. In those cases, the load-balancer is created with the user-specified loadBalancerIP. If the loadBalancerIP field is not specified, the loadBalancer is set up with an ephemeral IP address. If you specify a loadBalancerIP but your cloud provider does not support the feature, the loadbalancerIP field that you set is ignored.

**Release Note**

```release-note
NONE
```

### CODE REVIEW

**Comment by theomessin** [comment]
Example config after this Pull Request is merged which would automagically get Knative Serving working with Auto-TLS in Google Kubernetes Engine as long as the set IP address has been created in GCP.

```yaml
---
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  ingress:
    kourier:
      enabled: true
      service-loadBalancerIP: 1.2.3.4
  config:
    domain:
      "example.com": ""
    network:
      auto-tls: Enabled
      certificate-class: net-http01.certificate.networking.knative.dev
      http-protocol: Redirected
      ingress-class: kourier.ingress.networking.knative.dev
  additionalManifests:
    - URL: https://github.com/knative/net-http01/releases/download/knative-v1.9.0/release.yaml
```

**Review by nak3 — CHANGES_REQUESTED** [review]
Thank you for this PR @theomessin 

But I realized that the `LoadBalancerIP` field has been deprecated in k8s 1.24. Is it possible to ask your cloud provider to support the annotation as suggested in the change log?

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md#deprecation
> Deprecated Service.Spec.LoadBalancerIP. This field was under-specified and its meaning varies across implementations. As of Kubernetes v1.24, users are encouraged to use implementation-specific annotations when available. This field may be removed in a future API version.

**Comment by theomessin** [comment]
Hi @nak3. Yes, it has been marked as deprecated. We're using Google Kubernetes Engine and as far as I can tell they haven't implemented any alternatives as of yet. They still recommend we use `loadBalancerIP`. I'd say the best thing to do is still add this in the operator.

File: pkg/reconciler/knativeserving/ingress/kourier.go
Comment (nak3) [inline pkg/reconciler/knativeserving/ingress/kourier.go:112]: As per https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/
> Only applies to Service Type: LoadBalancer.

The `ServiceLoadBalancerIP` applies to service type `LoadBalancer` only and you also need to use it for `LoadBalancer`, right? Then, I would the add `if` condition in `configureGWServiceType()` to insert the logic to apply `ServiceLoadBalancerIP` like:

```go
                        case v1.ServiceTypeClusterIP, v1.ServiceTypeNodePort, v1.ServiceTypeLoadBalancer:
                                svc.Spec.Type = serviceType
                                if serviceType == v1.ServiceTypeNodePort && instance.Spec.Ingress.Kourier.ServiceLoadBalancerIP != ""{
                                        ...
                                }
```

Probably we should rename the function `configureGWServiceType()`  with `configureGWService()` then.

File: config/crd/bases/operator.knative.dev_knativeservings.yaml
Comment (nak3) [inline config/crd/bases/operator.knative.dev_knativeservings.yaml:2302]: `loadBalancerIP` is same style with k8s's but camel case and kebab case mixed.
@houshengbo Do you have any suggestion for the config name - like `service-loadBalancerIP`, `service-load-balancer-ip` or somethinge else?

Comment (theomessin) [reply config/crd/bases/operator.knative.dev_knativeservings.yaml:2314]: I'm happy to change this to `service-load-balancer-ip` if that's better. Let me know.

File: pkg/reconciler/knativeserving/ingress/kourier.go
Comment (theomessin) [reply pkg/reconciler/knativeserving/ingress/kourier.go]: It's not that simple. The default type for the `kourier` Service is `LoadBalancer`
https://github.com/knative/operator/blob/a3758e7915b60ea859cbc4951101708d92097302/cmd/operator/kodata/ingress/1.10/kourier/kourier.yaml#L576

If we do this, we'd have to change this early return check because a user might set `service-loadBalancerIP` but not `service-type`.
https://github.com/knative/operator/blob/037957ad5d3ccd22b12826a0a17f22b122d78a32/pkg/reconciler/knativeserving/ingress/kourier.go#L82-L84

I ended up refactoring the whole function. Let me know if you prefer this or if I should revert.

**Comment by theomessin** [comment]
@nak3 any update on this?

File: config/crd/bases/operator.knative.dev_knativeservings.yaml
Comment (nak3) [reply config/crd/bases/operator.knative.dev_knativeservings.yaml:2314]: I am prefer to `service-load-balancer-ip`. WDYT @houshengbo ?

**Review by nak3 — COMMENTED** [review]
Thank you! LGTM except for one comment.

File: config/crd/bases/operator.knative.dev_knativeservings.yaml
Comment (theomessin) [reply config/crd/bases/operator.knative.dev_knativeservings.yaml]: Done

**Comment by nak3** [comment]
/lgtm
/approve

/assign @houshengbo 
Thank you! I think I don't have a permission to merge this so defer to @houshengbo

**Comment by nak3** [comment]
bump @houshengbo
/cc @dprotaso @psschwei

**Comment by theomessin** [comment]
Are we only waiting on the lint failing? The error message doesn't mean anything to me...

```
Running [/home/runner/golangci-lint-1.52.0-linux-amd64/golangci-lint run --out-format=github-actions --new-from-patch=/tmp/tmp-1720-r08HIklCaYvx/pull.patch --new=false --new-from-rev= --go=1.20] in [] ...
  Error: unnecessary conversion (unconvert)
  ```

**Comment by theomessin** [comment]
@nak3 @houshengbo

File: pkg/reconciler/knativeserving/ingress/kourier_test.go
Comment (nak3) [inline pkg/reconciler/knativeserving/ingress/kourier_test.go:189]:
```suggestion
		util.AssertDeepEqual(t, svcLoadBalancerIP, expServiceLoadBalancerIP)
```

Could you fix the lint error?

**Comment by nak3** [comment]
As I commented https://github.com/knative/operator/pull/1343#issuecomment-1588287110, I don't have a power to approve PR.

**Comment by nak3** [comment]
/lgtm
/approve

(again I think I don't have a power to merge..)

**Review by psschwei — COMMENTED** [review]
I don't know the operator but trust @nak3 's judgment 
/approve
/hold
In case @houshengbo wants to weigh in, but feel free to unhold in a day or two

**Comment by theomessin** [comment]
@nak3 @psschwei @houshengbo Can we please get this merged? The longer it sits the more work I have to keep doing to resolve conflicts every time main changes...

**Pull request merged**
